src/modules/job-manager/queue.c

```diff
@@ -774,6 +776,171 @@ void queue_destroy (struct queue *queue)
     }
 }
 
+/*  Test equality of two constraint objects.
+ *  For now, two constraints are equivalent if:
+ *
+ *  - both are either NULL or empty objects (i.e. size == 0)
+ *    (Note: json_object_size (NULL) == 0)
+ *
+ *  - json_equal(a, b) returns true
+ */
+static bool constraints_equal (json_t *c1, json_t *c2)
+{
+    if ((json_object_size (c1) == 0 && json_object_size (c2) == 0)
+        || json_equal (c1, c2))
+        return true;
+    return false;
```

Comment:
I don't think you have to compare the two object sizes since [json_equal()](https://jansson.readthedocs.io/en/latest/apiref.html#c.json_equal) already requires that they have the same keys, per the jansson  [equality](https://jansson.readthedocs.io/en/latest/apiref.html#equality) definition.

Reply:
Oh I just realized I misread that line of code and you're checking for empty.  Well I would assume two empty objects would be considered equal by `json_equal()` but maybe you've actually found that not to be the case!

Reply:
No, the issue is that `json_object_size()` handles `NULL` (returns 0) and `json_equal()` does not (IIRC). We could check `c1 == NULL && c2 == NULL` but then empty objects would be unnecessarily passed to `json_equal()` for comparison, it seemed better to do that in one go... However, if it is confusing maybe the former is better code!

Reply:
Just verified if either argument is `NULL` then `json_equal()` will return false, however here we want `NULL == NULL` to return true. Maybe just explicitly testing for `NULL == NULL` would be least confusing...